File: samples/cpp/tutorial_icc_color_management.cpp
/**
 * @file icc_color_management_tutorial.cpp
 * @brief Tutorial demonstrating ICC v5/iccMAX color management in OpenCV
 *
 * This tutorial shows how to use the new ICC color management features:
 * - Loading and using ICC profiles
 * - Converting between color spaces with ICC profiles
 * - Using color appearance models (CAM02/CAM16)
 * - HDR and wide-gamut workflows
 */

#include <opencv2/opencv.hpp>
#include <iostream>

int main()
{
    std::cout << "OpenCV ICC Color Management Tutorial\n";
    std::cout << "====================================\n\n";

    try {
        // 1. Create test image (HDR content)
        cv::Mat hdrImage(300, 400, CV_32FC3);

        // Generate HDR gradient with values > 1.0
        for (int y = 0; y < hdrImage.rows; y++) {
            for (int x = 0; x < hdrImage.cols; x++) {
                cv::Vec3f& pixel = hdrImage.at<cv::Vec3f>(y, x);
                pixel[0] = (float)x / hdrImage.cols * 2.0f;  // Blue: 0-2.0
                pixel[1] = (float)y / hdrImage.rows * 3.0f;  // Green: 0-3.0
                pixel[2] = 1.5f;                             // Red: constant 1.5
            }
        }

        std::cout << "1. Created HDR test image (" << hdrImage.rows << "x" << hdrImage.cols << ")\n";
        std::cout << "   - Floating-point values: 0.0 to 3.0 (extends beyond SDR range)\n\n";

        // 2. Create ICC profiles for different color spaces
        cv::IccProfile srgbProfile = cv::createStandardProfile("sRGB");
        cv::IccProfile adobeRgbProfile = cv::createStandardProfile("Adobe RGB");
        cv::IccProfile proPhotoProfile = cv::createStandardProfile("ProPhoto RGB");
        cv::IccProfile rec2020Profile = cv::createStandardProfile("Rec2020");

        std::cout << "2. Created ICC profiles:\n";
        std::cout << "   - sRGB: " << srgbProfile.getDescription() << "\n";
        std::cout << "   - Adobe RGB: " << adobeRgbProfile.getDescription() << "\n";
        std::cout << "   - ProPhoto RGB: " << proPhotoProfile.getDescription() << "\n";
        std::cout << "   - Rec2020: " << rec2020Profile.getDescription() << "\n\n";

        // 3. Set up viewing conditions for color appearance modeling
        cv::ViewingConditions officeConditions = cv::getStandardViewingConditions("office");
        cv::ViewingConditions cinemaConditions = cv::getStandardViewingConditions("cinema");

        std::cout << "3. Viewing conditions:\n";
        std::cout << "   - Office: " << officeConditions.adaptingLuminance << " cd/m²\n";
        std::cout << "   - Cinema: " << cinemaConditions.adaptingLuminance << " cd/m²\n\n";

        // 4. Basic color space conversion: sRGB -> Adobe RGB
        cv::Mat adobeRgbResult;
        cv::colorProfileTransform(hdrImage, adobeRgbResult,
                                 srgbProfile, adobeRgbProfile,
                                 cv::ICC_PERCEPTUAL);

        std::cout << "4. Basic transformation: sRGB -> Adobe RGB\n";
        std::cout << "   - Rendering intent: Perceptual\n";
        std::cout << "   - Result size: " << adobeRgbResult.size() << "\n\n";

        // 5. Wide-gamut transformation: sRGB -> ProPhoto RGB
        cv::Mat proPhotoResult;
        cv::colorProfileTransform(hdrImage, proPhotoResult,
                                 srgbProfile, proPhotoProfile,
                                 cv::ICC_RELATIVE_COLORIMETRIC);

        std::cout << "5. Wide-gamut transformation: sRGB -> ProPhoto RGB\n";
        std::cout << "   - Rendering intent: Relative Colorimetric\n";
        std::cout << "   - Preserves HDR values and wide color gamut\n\n";

        // 6. HDR workflow: sRGB -> Rec2020 with CAM16 color appearance model
        cv::Mat rec2020Result;
        cv::colorProfileTransform(hdrImage, rec2020Result,
                                 srgbProfile, rec2020Profile,
                                 cv::ICC_PERCEPTUAL, cv::CAM16,
                                 officeConditions);

        std::cout << "6. HDR workflow: sRGB -> Rec2020 with CAM16\n";
        std::cout << "   - Color appearance model: CAM16\n";
        std::cout << "   - Viewing conditions: Office lighting\n";
        std::cout << "   - Accurate perceptual color adaptation\n\n";

        // 7. Compare different rendering intents
        cv::Mat perceptualResult, colorimetricResult, saturationResult;

        cv::colorProfileTransform(hdrImage, perceptualResult,
                                 srgbProfile, adobeRgbProfile,
                                 cv::ICC_PERCEPTUAL);

        cv::colorProfileTransform(hdrImage, colorimetricResult,
                                 srgbProfile, adobeRgbProfile,
                                 cv::ICC_RELATIVE_COLORIMETRIC);

        cv::colorProfileTransform(hdrImage, saturationResult,
                                 srgbProfile, adobeRgbProfile,
                                 cv::ICC_SATURATION);

        std::cout << "7. Rendering intent comparison:\n";
        std::cout << "   - Perceptual: Optimized for photographic content\n";
        std::cout << "   - Relative Colorimetric: Preserves color accuracy\n";
        std::cout << "   - Saturation: Enhances vivid colors\n\n";

        // 8. Single color transformation (useful for UI/graphics)
        cv::Mat singleColor = (cv::Mat_<float>(1, 3) << 0.8f, 1.2f, 2.1f); // HDR color
        cv::Mat transformedColor;

        cv::colorProfileTransformSingle(singleColor, transformedColor,
                                       srgbProfile, proPhotoProfile,
                                       cv::ICC_PERCEPTUAL, cv::CAM02,
                                       officeConditions);

        std::cout << "8. Single color transformation:\n";
        std::cout << "   - Input HDR color: [" << singleColor.at<float>(0,0)
                  << ", " << singleColor.at<float>(0,1)
                  << ", " << singleColor.at<float>(0,2) << "]\n";
        std::cout << "   - Transformed with CIECAM02 appearance model\n\n";

        // 9. Professional workflow demonstration
        std::cout << "9. Professional Workflow Examples:\n\n";

        // Cinema workflow: Rec2020 -> sRGB with cinema viewing conditions
        cv::Mat cinemaToSrgb;
        cv::colorProfileTransform(hdrImage, cinemaToSrgb,
                                 rec2020Profile, srgbProfile,
                                 cv::ICC_PERCEPTUAL, cv::CAM16,
                                 cinemaConditions);
        std::cout << "   • Cinema mastering: Rec2020 -> sRGB with cinema conditions\n";

        // Photography workflow: Adobe RGB -> ProPhoto RGB
        cv::Mat photoWorkflow;
        cv::colorProfileTransform(hdrImage, photoWorkflow,
                                 adobeRgbProfile, proPhotoProfile,
                                 cv::ICC_RELATIVE_COLORIMETRIC);
        std::cout << "   • Photography: Adobe RGB -> ProPhoto RGB (archival)\n";

        // Print workflow: sRGB -> CMYK simulation (using Adobe RGB as proxy)
        cv::Mat printWorkflow;
        cv::colorProfileTransform(hdrImage, printWorkflow,
                                 srgbProfile, adobeRgbProfile,
                                 cv::ICC_PERCEPTUAL, cv::CAM_NONE,
                                 cv::getStandardViewingConditions("print"));
        std::cout << "   • Print preview: sRGB -> print simulation with print conditions\n\n";

        // 10. Performance and compatibility notes
        std::cout << "10. Performance Notes:\n";
        std::cout << "    • Use floating-point formats (CV_32F, CV_64F) for HDR content\n";
        std::cout << "    • CAM02/CAM16 models add computational overhead but improve accuracy\n";
        std::cout << "    • Relative colorimetric intent preserves out-of-gamut HDR values\n";
        std::cout << "    • Multi-threading supported for large images\n\n";

        std::cout << "Tutorial completed successfully!\n";
        std::cout << "ICC v5/iccMAX color management provides:\n";
        std::cout << "• True HDR and wide-gamut support\n";
        std::cout << "• Perceptual color appearance modeling\n";
        std::cout << "• Professional cinema/photography workflows\n";
        std::cout << "• Backward compatibility with existing code\n";

    } catch (const cv::Exception& e) {
        std::cerr << "OpenCV Error: " << e.what() << std::endl;
        return -1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return -1;
    }

    return 0;
}

File: samples/python/tutorial_icc_color_management.py
#!/usr/bin/env python3
"""
OpenCV ICC Color Management Tutorial (Python)
============================================

This tutorial demonstrates the new ICC v5/iccMAX color management features
in OpenCV Python bindings, including:
- HDR and wide-gamut color workflows
- Color appearance models (CAM02/CAM16)
- Professional cinema and photography pipelines
- Multi-channel and spectral color processing
"""

import cv2
import numpy as np

def main():
    print("OpenCV ICC Color Management Tutorial (Python)")
    print("=" * 50)
    print()

    try:
        # 1. Create HDR test image with extended dynamic range
        height, width = 300, 400
        hdr_image = np.zeros((height, width, 3), dtype=np.float32)

        # Generate HDR gradient (values > 1.0)
        for y in range(height):
            for x in range(width):
                hdr_image[y, x, 0] = (x / width) * 2.0    # Blue: 0-2.0
                hdr_image[y, x, 1] = (y / height) * 3.0   # Green: 0-3.0
                hdr_image[y, x, 2] = 1.5                  # Red: constant 1.5

        print(f"1. Created HDR test image ({height}x{width})")
        print(f"   - Floating-point values: 0.0 to 3.0 (extends beyond SDR)")
        print(f"   - Min value: {hdr_image.min():.2f}, Max value: {hdr_image.max():.2f}")
        print()

        # 2. Create ICC profiles for different color spaces
        srgb_profile = cv2.createStandardProfile("sRGB")
        adobe_rgb_profile = cv2.createStandardProfile("Adobe RGB")
        prophoto_profile = cv2.createStandardProfile("ProPhoto RGB")
        rec2020_profile = cv2.createStandardProfile("Rec2020")

        print("2. Created ICC profiles:")
        print(f"   - sRGB: {srgb_profile.getDescription()}")
        print(f"   - Adobe RGB: {adobe_rgb_profile.getDescription()}")
        print(f"   - ProPhoto RGB: {prophoto_profile.getDescription()}")
        print(f"   - Rec2020: {rec2020_profile.getDescription()}")
        print()

        # 3. Set up viewing conditions
        office_conditions = cv2.getStandardViewingConditions("office")
        cinema_conditions = cv2.getStandardViewingConditions("cinema")
        print_conditions = cv2.getStandardViewingConditions("print")

        print("3. Viewing conditions:")
        print(f"   - Office: {office_conditions.adaptingLuminance} cd/m²")
        print(f"   - Cinema: {cinema_conditions.adaptingLuminance} cd/m²")
        print(f"   - Print: {print_conditions.adaptingLuminance} cd/m²")
        print()

        # 4. Basic color space conversion
        adobe_result = cv2.colorProfileTransform(
            hdr_image, srgb_profile, adobe_rgb_profile,
            intent=cv2.ICC_PERCEPTUAL
        )

        print("4. Basic transformation: sRGB → Adobe RGB")
        print(f"   - Rendering intent: Perceptual")
        print(f"   - Result shape: {adobe_result.shape}")
        print(f"   - Value range: {adobe_result.min():.2f} to {adobe_result.max():.2f}")
        print()

        # 5. Wide-gamut HDR workflow
        prophoto_result = cv2.colorProfileTransform(
            hdr_image, srgb_profile, prophoto_profile,
            intent=cv2.ICC_RELATIVE_COLORIMETRIC
        )

        print("5. Wide-gamut transformation: sRGB → ProPhoto RGB")
        print(f"   - Preserves HDR values and wide color gamut")
        print(f"   - Value range: {prophoto_result.min():.2f} to {prophoto_result.max():.2f}")
        print()

        # 6. Color appearance model workflow
        rec2020_cam16 = cv2.colorProfileTransform(
            hdr_image, srgb_profile, rec2020_profile,
            intent=cv2.ICC_PERCEPTUAL,
            cam=cv2.CAM16,
            viewingConditions=office_conditions
        )

        print("6. HDR workflow with CAM16: sRGB → Rec2020")
        print(f"   - Color appearance model: CAM16")
        print(f"   - Viewing conditions: Office lighting")
        print(f"   - Perceptual adaptation applied")
        print()

        # 7. Compare different rendering intents
        intents = {
            'Perceptual': cv2.ICC_PERCEPTUAL,
            'Relative Colorimetric': cv2.ICC_RELATIVE_COLORIMETRIC,
            'Saturation': cv2.ICC_SATURATION,
            'Absolute Colorimetric': cv2.ICC_ABSOLUTE_COLORIMETRIC
        }

        print("7. Rendering intent comparison:")
        results = {}
        for name, intent in intents.items():
            result = cv2.colorProfileTransform(
                hdr_image, srgb_profile, adobe_rgb_profile, intent=intent
            )
            results[name] = result
            print(f"   - {name}: Range {result.min():.2f} to {result.max():.2f}")
        print()

        # 8. Single color transformation
        hdr_color = np.array([[[0.8, 1.2, 2.1]]], dtype=np.float32)
        transformed_color = cv2.colorProfileTransformSingle(
            hdr_color, srgb_profile, prophoto_profile,
            intent=cv2.ICC_PERCEPTUAL,
            cam=cv2.CAM02,
            viewingConditions=office_conditions
        )

        print("8. Single color transformation:")
        print(f"   - Input HDR color: {hdr_color[0,0]}")
        print(f"   - Transformed color: {transformed_color[0,0]}")
        print(f"   - Used CIECAM02 appearance model")
        print()

        # 9. Professional workflows
        print("9. Professional Workflow Examples:")
        print()

        # Cinema mastering workflow
        cinema_master = cv2.colorProfileTransform(
            hdr_image, rec2020_profile, srgb_profile,
            intent=cv2.ICC_PERCEPTUAL,
            cam=cv2.CAM16,
            viewingConditions=cinema_conditions
        )
        print("   • Cinema mastering: Rec2020 → sRGB with cinema viewing")

        # Photography workflow
        photo_workflow = cv2.colorProfileTransform(
            hdr_image, adobe_rgb_profile, prophoto_profile,
            intent=cv2.ICC_RELATIVE_COLORIMETRIC
        )
        print("   • Photography: Adobe RGB → ProPhoto RGB (archival)")

        # Print simulation
        print_sim = cv2.colorProfileTransform(
            hdr_image, srgb_profile, adobe_rgb_profile,
            intent=cv2.ICC_PERCEPTUAL,
            viewingConditions=print_conditions
        )
        print("   • Print preview: sRGB → print simulation")
        print()

        # 10. Machine learning dataset preparation
        print("10. Machine Learning Applications:")

        # Create dataset with consistent color management
        dataset_colors = []
        source_spaces = [srgb_profile, adobe_rgb_profile, prophoto_profile]
        target_space = rec2020_profile  # Standardize to wide gamut

        for i, source_profile in enumerate(source_spaces):
            # Simulate different camera/source color spaces
            sample_image = hdr_image * (0.8 + i * 0.1)  # Vary exposure

            standardized = cv2.colorProfileTransform(
                sample_image, source_profile, target_space,
                intent=cv2.ICC_RELATIVE_COLORIMETRIC,
                cam=cv2.CAM16,
                viewingConditions=office_conditions
            )

            dataset_colors.append(standardized)
            print(f"    • Standardized {source_profile.getColorSpace()} → Rec2020")

        print("    • Consistent color representation across training data")
        print("    • Improved model generalization with accurate colors")
        print()

        # 11. Performance optimization tips
        print("11. Performance Optimization:")
        print("    • Use numpy arrays with dtype=np.float32 for HDR content")
        print("    • Batch process multiple images for efficiency")
        print("    • Cache ICC profiles to avoid repeated parsing")
        print("    • Use appropriate rendering intent for your use case")
        print("    • Consider computational cost of color appearance models")
        print()

        # 12. Integration with existing OpenCV workflows
        print("12. Integration Examples:")

        # Traditional cvtColor still works
        srgb_traditional = cv2.cvtColor(hdr_image.astype(np.uint8), cv2.COLOR_BGR2RGB)
        print("    • Traditional cvtColor: Compatible with existing code")

        # ICC-aware processing
        icc_aware = cv2.colorProfileTransform(
            hdr_image, srgb_profile, adobe_rgb_profile
        )
        print("    • ICC-aware processing: Enhanced color accuracy")

        # Hybrid workflow
        gray = cv2.cvtColor(icc_aware, cv2.COLOR_BGR2GRAY)
        print("    • Hybrid: ICC transform → traditional processing")
        print()

        print("Tutorial completed successfully!")
        print()
        print("OpenCV ICC v5/iccMAX enables:")
        print("✓ True HDR and wide-gamut color processing")
        print("✓ Perceptual color appearance modeling")
        print("✓ Professional cinema/photography workflows")
        print("✓ Improved machine learning dataset quality")
        print("✓ Seamless integration with existing OpenCV code")

    except Exception as e:
        print(f"Error: {e}")
        return -1

    return 0

if __name__ == "__main__":
    main()
